### PR TITLE
Parse syntax trees in parallel when producing initial compilations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 1]
```diff
@@ -324,6 +324,15 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
                         : this.ProjectState;
 
                     compilationWithoutGeneratedDocuments = inProgressState.CompilationWithoutGeneratedDocuments;
+
+                    // Parse the generated documents in parallel.
+                    Parallel.ForEach(
+                        generatorInfo.Documents.States.Values,
+                        new ParallelOptions { CancellationToken = cancellationToken },
+                        (state, x) => state.GetSyntaxTree(cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:332]:
```suggestion
                        state => state.GetSyntaxTree(cancellationToken));
```

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs [hunk 2]
```diff
@@ -93,10 +93,8 @@ public TState GetRequiredState(DocumentId documentId)
         /// <returns></returns>
         public IEnumerable<TState> GetStatesInCompilationOrder()
         {
-            foreach (var id in Ids)
-            {
-                yield return _map[id];
-            }
+            var map = _map;
+            return Ids.Select(id => map[id]);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs:97]: intentional chance, this allows the runtime to produce an IEnumerable with a known length, which helps with any operations it does where it wants to allocate space.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs [hunk 3]
```diff
@@ -93,10 +93,8 @@ public TState GetRequiredState(DocumentId documentId)
         /// <returns></returns>
         public IEnumerable<TState> GetStatesInCompilationOrder()
         {
-            foreach (var id in Ids)
-            {
-                yield return _map[id];
-            }
+            var map = _map;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs:96]: required as we have a lambda capturing within a struct.

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 4]
```diff
@@ -324,6 +324,15 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
                         : this.ProjectState;
 
                     compilationWithoutGeneratedDocuments = inProgressState.CompilationWithoutGeneratedDocuments;
+
+                    // Parse the generated documents in parallel.
+                    Parallel.ForEach(
+                        generatorInfo.Documents.States.Values,
+                        new ParallelOptions { CancellationToken = cancellationToken },
+                        state => state.GetSyntaxTree(cancellationToken));
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:332]: This worries me if potentially Parallel.ForEach() were to try to start running stuff on other threads and end up getting blocked there. But if the framework is smart enough to keep using our own thread which potentially might do all the work then this isn't any worse than it is now.

You hit a more general point that it seems we're actually parsing files here, so I wonder if this another place we can defer this work until later like we did for the regular document. But you've just highlighted a problem I didn't appreciate until now.

**Review by jasonmalinowski — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:323]: > FreezePartialState

yup.  that's one of the things i was exploring.

File: src/EditorFeatures/Test2/Rename/RenameCommandHandlerTests.vb [hunk 5]
```diff
@@ -24,7 +24,7 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.Rename
 
         <WpfTheory>
         <CombinatorialData, Trait(Traits.Feature, Traits.Features.Rename)>
-        Public Sub RenameCommandInvokesInlineRename(host As RenameTestHost)
+        Public Async Sub RenameCommandInvokesInlineRename(host As RenameTestHost)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Rename/RenameCommandHandlerTests.vb:27]:
```suggestion
        Public Async Function RenameCommandInvokesInlineRename(host As RenameTestHost)
```

**Pull request merged**